docs/P4c-test-documentation.md

```diff
@@ -0,0 +1,235 @@
+# **Adding New Tests to the P4C Repository**
+
+**This document provides a guide for developers who want to add new tests to the P4C repository. It explains where to place test files, how to manage expected output files, and how to handle tests that are expected to fail (Xfail).**
+
+---
+
+
+
+
+
+## **1. Adding "Positive" Tests**
+
+Positive tests are those where the P4 compiler should process the input without any errors.
+
+### Steps for Adding Positive Tests
+
+1. **Source File Location**:
+
+   - Positive tests are located in multiple directories based on the purpose and the backend/tool being tested. Below is a description of the primary directories under `p4c/testdata` and their purposes:
+
+     - **`p4_16_samples`**: Contains general positive tests for the P4\_16 language.
+     - **`p4_16_samples_outputs`**: Stores the expected outputs for the tests in `p4_16_samples`.
+     - **`p4_14_samples`**: Contains tests specific to the older P4\_14 language version.
+     - **`p4_14_samples_outputs`**: Stores the expected outputs for tests in `p4_14_samples`.
+     - **`bmv2`**: Holds tests targeting the BMv2 backend, including packet-processing logic and examples.
+     - **`psa`**: Includes tests for the PSA (Portable Switch Architecture) target.
+     - **`ebpf`**: Contains tests designed for the eBPF backend.
+     - **`xdp`**: Stores tests related to XDP programs using eBPF.
+     - **`arch`**: Contains architecture-specific tests, covering various P4 architectures.
+     - **`tools`**: Includes tests for tools provided in the P4C repository, such as `p4test` and `p4c-bm2-ss`.
```

Comment:
I can find a `tools` directory, but it does not contain individual unit tests for `p4test` and `p4c-bm2-ss`.  It contains what I would call "test driver programs" (maybe there is a different name used somewhere in existing documentation), that are used by many, many test programs.  While it would be nice to mention them here, true, for the knowledge of p4c developers, most developers adding new test programs will not be modifying the files in the `tools` directory.

They _will_ commonly be adding files to subdirectories under the `p4c/testdata` directory, and I would hope that list of directories should all be mentioned in this list, and what they are for.